worker/provisioner/provisioner.go

```diff
@@ -86,18 +86,27 @@ func NewRetryStrategy(delay time.Duration, count int) RetryStrategy {
 	}
 }
 
-// configObserver is implemented so that tests can see
-// when the environment configuration changes.
+// configObserver is implemented so that tests can see when the environment
+// configuration changes.
+// The dying function should be set by the outer provider to return its
+// catacomb. This is used to prevent notify from blocking a provisioner
+// that has had its Kill method invoked.
 type configObserver struct {
 	sync.Mutex
 	observer chan<- *config.Config
+	dying    func() <-chan struct{}
 }
 
 // notify notifies the observer of a configuration change.
 func (o *configObserver) notify(cfg *config.Config) {
 	o.Lock()
 	if o.observer != nil {
-		o.observer <- cfg
+		if o.observer != nil {
```

Comment:
I don't think you need the double nested o.observer != nil.

I'm guessing you were thinking about turning it into a for loop? But I think since its either "we send the message, or we die" we probably don't need any sort of retry.

Reply:
Oops. Fixing.